src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1420,16 +1430,75 @@ private IEnumerable<TypeInferenceInfo> InferTypeInImplicitArrayCreation(Implicit
                 return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
             }
 
+            private IEnumerable<TypeInferenceInfo> InferTypeInPropertyPattern(PropertyPatternSyntax propertyPattern)
+            {
+                // something like: { ... sub patterns ... } or like: Blah { ... sub patterns ... }
+                // in the latter case, we know the type, because it's explicitly stated.  In the former
+                // we have to walk higher to figure it out.
+                var type = this.SemanticModel.GetTypeInfo(propertyPattern).ConvertedType;
+                return type == null
+                    ? SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>()
+                    : SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(type));
+            }
+
+            private IEnumerable<TypeInferenceInfo> InferTypeInConstantPattern(
+                ConstantPatternSyntax constantPattern)
+            {
+                return InferTypes(constantPattern);
+            }
+
+            private IEnumerable<TypeInferenceInfo> InferTypeInPropertySubpattern(
+                PropertySubpatternSyntax propertySubpattern,
+                SyntaxNode child)
+            {
+                return InferTypes(propertySubpattern);
+            }
+
+            private IEnumerable<TypeInferenceInfo> InferTypeInSubpatternElement(
```

Comment:
I assume this inference comes into play when typing `{ Property: SomeNewMethod() }` and triggering "generate method" on `SomeNewMethod`. If that's the case, could you add a test like that.

Along those lines, it feels that most of the changes to type inferrer aren't covered by tests.

Reply:
@jcouv I don't understand. In what case can an invocation be a constant expression in order to appear in a pattern?

Reply:
There is a test case for this.   Note: it's not an invocation per-se, just a simple name reference.  But the inference code comes into play.

Reply:
Hum, good point. Then I'd think we can remove this method ;-)

Reply:
Oh, I see. When generating a variable in `{ [|X|] : int i }`, you need to infer the type of `int i` to give the generated "X" a type.
Makes sense

Reply:
@jcouv I don't think so. Why would we use a type pattern for `int i` if we know that the original type is an `int`? That doesn't make sense to me. In the majority of cases when we use patterns, this type would be different.

Reply:
Yes.  The purpose of the inference engine is to use what's *around* the code to help determine what we should thinka bout the node in question.  That can happen in multiple directions.  So, for example ```[|X|]: Y```, we woudl try to figure out the type of X based on the tyep of Y.  Conversely if you have ```X: [|Y|]```, we'd try to figure out the tyep of Y based on the type of X.

Reply:
> @jcouv I don't think so. Why would we use a type pattern for int i if we know that the original type is an int? 

Because it's just something reasonable to do.  Yes, the user may end up not wanting that.  But maybe they just wanted this to give a name to the variable in deconstruction and they liked being explicit.

Another example would be ```X : string x``` here you want it to be 'string' with the user likely havin this pattern to match when non-null.

Reply:
I believe the appropriate pattern for non-null is `{ }` without the type.

Reply:
That is the pattern for when people do not want to repeat the type.  But some people want to specify the type along with their variables :)

Reply:
```{}``` is intended to be the counterpart for the *var* pattern.  i.e. you want it ot match all the time, except when non-null.  So it's useful for people who like using 'var' but don't like that 'var' matched null.